ops/testing.py

```diff
@@ -403,9 +404,84 @@ def add_storage(self, storage_name: str, count: int = 1) -> int:
         """
-        return self._backend.storage_add(storage_name, count)
+        if storage_name not in self._meta.storages:
+            raise RuntimeError(
+                "the key '{}' is not specified as a storage key in metadata".format(storage_name))
+        storage_indices = self._backend.storage_add(storage_name, count)
+
+        # Reset associated cached value in the storage mappings.  If we don't do this,
+        # Model._storages won't return Storage objects for subsequently-added storage.
+        self._model._storages._storage_map[storage_name] = None
+
+        if self.charm is not None and self._hooks_enabled:
+            for storage_index in storage_indices:
+                self.charm.on[storage_name].storage_attached.emit(
+                    model.Storage(storage_name, storage_index, self._backend))
+        return ["{}/{}".format(storage_name, storage_index) for storage_index in storage_indices]
+
+    def detach_storage(self, storage_id: str) -> None:
+        """Detach a storage device.
+
+        The intent of this function is to simulate a "juju detach-storage" call.
+        It will trigger a storage-detaching hook if the storage unit in question exists
+        and is presently marked as attached.
+
+        Args:
+            storage_id: The full storage ID of the storage unit being detached, including the
+                storage key, e.g. my-storage/0.
+        """
+        if self.charm is None:
+            raise RuntimeError('cannot detach storage before Harness is initialised')
+        storage_name, storage_index = storage_id.split('/', 1)
+        storage_index = int(storage_index)
+        if self._backend._storage_is_attached(storage_name, storage_index) and self._hooks_enabled:
+            self.charm.on[storage_name].storage_detaching.emit(
+                model.Storage(storage_name, storage_index, self._backend))
+        self._backend._storage_detach(storage_id)
+
+    def attach_storage(self, storage_id: str) -> None:
+        """Attach a storage device.
+
+        The intent of this function is to simulate a "juju attach-storage" call.
+        It will trigger a storage-attached hook if the storage unit in question exists
+        and is presently marked as detached.
+
+        Args:
+            storage_id: The full storage ID of the storage unit being attached, including the
+                storage key, e.g. my-storage/0.
+        """
+        if self.charm is None:
+            raise RuntimeError('cannot attach storage before Harness is initialised')
+        if self._backend._storage_attach(storage_id) and self._hooks_enabled:
+            storage_name, storage_index = storage_id.split('/', 1)
+            storage_index = int(storage_index)
+            self.charm.on[storage_name].storage_attached.emit(
+                model.Storage(storage_name, storage_index, self._backend))
+
+    def remove_storage(self, storage_id: str) -> None:
```

Comment:
If the charm can't see it, I think that it shouldn't exist in the testing harness.

Caveat: if it is somehow possible to remove storage without detaching it first, then it would be useful to be able to test charm code that handles that case. I think that would count as a Juju bug, though, and charm authors probably shouldn't have to worry about it ...

Reply:
If there's no real difference between removed vs detached or created vs attached at the level of the charm itself, then sure, I'm in favor of removing the attach/detach bits.

Reply:
Yeh, I can see how this might simplify things slightly. Ultimately so long as we have a way for charmers to say "hey charm, here's some storage" and validate that their hooks do the right thing, then we're good - certainly for a first implementation.

Reply:
Per discussions on Mattermost, there are some older tests that rely on add/remove, not just attach/detach. We're going to leave the methods in for now.